aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -200,16 +200,23 @@ TORCH_META_FUNC2(prod, dim_int)
   resize_reduction(*this, self, dim, keepdim, out_dtype);
 }
 
-void check_floating_or_complex_dtype(const char* name, ScalarType dtype) {
-  TORCH_CHECK(
-      at::isFloatingType(dtype) || at::isComplexType(dtype),
-      name, "(): input dtype should be either floating point or complex dtypes. "
-      "Got ", toString(dtype), " instead.");
-}
-
 TORCH_META_FUNC2(mean, dim)
 (const Tensor& self, IntArrayRef dim, bool keepdim, optional<ScalarType> opt_dtype) {
-  check_floating_or_complex_dtype("mean", self.scalar_type());
+  auto in_dtype = at::native::get_dtype_from_self(self, opt_dtype, true);
+
+  if (!at::isFloatingType(in_dtype) && !at::isComplexType(in_dtype)) {
+    std::string opt_dtype_str("None");
+    if (opt_dtype.has_value()) {
+      opt_dtype_str = toString(opt_dtype.value());
+    }
+
+    TORCH_CHECK(
+        false,
+        "mean(): at least one of (i) the input dtype and (ii) the desired output dtype should be "
```

Comment:
this is a misleading error message, e.g. if self dtype if float but dtype arg is integer, then it's and invalid combination (and you'll rightly error out), but error message wouldn't explain it.